src/bin/sccache-dist/build.rs

```diff
@@ -190,12 +202,36 @@ impl OverlayBuilder {
                         )
                     }
                 };
-                tar::Archive::new(GzDecoder::new(toolchain_rdr)).unpack(&toolchain_dir)?;
+                tar::Archive::new(GzDecoder::new(toolchain_rdr))
+                    .unpack(&toolchain_dir)
+                    .or_else(|e| {
+                        warn!("Failed to unpack toolchain: {:?}", e);
+                        fs::remove_dir_all(&toolchain_dir)
+                            .chain_err(|| "Failed to remove unpacked toolchain")?;
+                        tccache
+                            .remove(tc)
+                            .chain_err(|| "Failed to remove corrupt toolchain")?;
+                        Err(Error::from(e))
+                    })?;
+
+                let entry = DeflatedToolchain {
+                    path: toolchain_dir,
+                    build_count: 1,
+                    ctime: Instant::now(),
+                };
 
-                let entry = (toolchain_dir, 1);
-                assert!(toolchain_dir_map
-                    .insert(tc.clone(), entry.clone())
-                    .is_none());
+                toolchain_dir_map.insert(tc.clone(), entry.clone());
+                if toolchain_dir_map.len() > tccache.len() {
+                    let dir_map = toolchain_dir_map.clone();
+                    let mut entries: Vec<_> = dir_map.iter().collect();
+                    entries.sort_by(|a, b| (a.1).ctime.cmp(&(b.1).ctime));
+                    if let Some((tc, _)) = entries.first() {
+                        warn!("Removing old un-compressed toolchain: {:?}", tc);
+                        assert!(toolchain_dir_map.remove(tc).is_some());
+                        fs::remove_dir_all(&self.dir.join("toolchains").join(&tc.archive_id))
+                            .chain_err(|| "Failed to remove old toolchain directory")?;
+                    }
+                }
```

Comment:
I think I would prefer if this were split out into its own pull request and not ride along with the corruption fixes.

I'm also not sure if this is the best way to do things.  IIUC, we have `toolchain_dir_map`, which maps from `Toolchain` to (effectively) some path.  And then we have the `tccache`, which is mapping from paths to stuff.  And this block is compensating for having this double set of mappings, in some sense?  Should we really have a unified mapping that also manages the LRU-ness of the underlying disk cache?  Or would that confuse other parts of the system that don't really expect to have toolchains disappear from underneath them?

Maybe all these questions are easily answerable, maybe not.  In any event, I think this should be a separate thing.

Reply:
I'm happy to split this out into more commits, but I'm not entirely keen on a deeper rewrite of these wonky caching layers. The current code is going to lead to servers running out of disk, and I think this will be a reasonably reliable way to prevent that in the short term. Do you have a specific approach in mind? The implementation is still quite weird, because we're allowing people to specify the max size of the compressed toolchain cache but leaving corresponding unpacked toolchains around on disk for quite a while. Perhaps a better approach would just be to cache unpacked toolchains rather than having a two tier system.